src/midgard/tiles.cc

```diff
@@ -17,25 +17,24 @@ namespace {
 // at each step it decides to either move in the x or y direction based on which pixels midpoint
 // forms a smaller triangle with the line. to avoid edge cases we allow set_pixel to make the
 // the loop bail if we leave the valid drawing region
-void bresenham_line(float x0,
-                    float y0,
-                    float x1,
-                    float y1,
+void bresenham_line(double x0,
+                    double y0,
+                    double x1,
+                    double y1,
                     const std::function<bool(int32_t, int32_t)>& set_pixel) {
   // this one for sure
   bool outside = set_pixel(std::floor(x0), std::floor(y0));
   // steps in the proper direction and constants for shoelace formula
-  float sx = x0 < x1 ? 1 : -1, dx = x1 - x0, x = std::floor(x0) + .5f;
-  float sy = y0 < y1 ? 1 : -1, dy = y1 - y0, y = std::floor(y0) + .5f;
+  double sx = x0 < x1 ? 1 : -1, dx = x1 - x0, x = std::floor(x0) + .5f;
+  double sy = y0 < y1 ? 1 : -1, dy = y1 - y0, y = std::floor(y0) + .5f;
   // keep going until we make it to the ending pixel
   while (std::floor(x) != std::floor(x1) || std::floor(y) != std::floor(y1)) {
-    float tx = std::abs(dx * (y - y0) - dy * ((x + sx) - x0));
-    float ty = std::abs(dx * ((y + sy) - y0) - dy * (x - x0));
+    double tx = std::abs(dx * (y - y0) - dy * ((x + sx) - x0));
+    double ty = std::abs(dx * ((y + sy) - y0) - dy * (x - x0));
     // less error moving in the x
-    if (tx < ty) {
+    if (tx < ty || (tx == ty && y0 == y1)) {
```

Comment:
this is the extra layer of protection here. i put a bunch of comments on the two sets of new test cases i added that explains what is going on. TLDR; moving to double precision makes it so that this line here isnt needed for spherical coordinates (which is really the only thing we care about). this is because i have code below that before calling this function first resamples (adds more coordinates) along the geodesic between two polar coordinates in an edge shape. in practice, when doing this it makes it so that this bresenham method will never be called on a fully horizontal set of points. however i kept the fix because for planer (cartesian) coordinates we do not resample (since lines in the plane dont curve).

the root problem was mixing double and float. in the (`Intersect`) method that calls this method, we check for some easy cases, if two points are in the same bin we mark the bin and exit, if they are adjacent we mark the two and exit. the shapes we found were mostly horiztonal but ever so slightly had differences in the y coordinate. so when you looked at them with double precision they would, for example when looking at a quad of 4 bins, start in the bottom left bin and end in the upper right. this is not one of the simple cases, so it needs to come to bresenham to be rasterized. however when it gets here and we do the math in single precision float it thinks that the line is moving only horizontally which causes this tie breaker where it defaults to moving in the y direction.